Change waitForBody to use one MutationObserver for N calls

choumx opened the pull request:
Fixes #4040. Instead of instantiating N MutationObservers for N calls to waitForBody, instantiate one and use an Observable in DocumentState to store the event handlers.

Seems to offer a modest performance improvement for large N. Tested rough approach on desktop Chrome:
N = 10000, 127.7 vs 5.2 ms (25x faster)
N = 1000, 4.14 vs 0.37 ms (11x faster)
N = 100, 0.52 vs 0.25 ms (2x faster)
N = 10, 0.26 vs 0.27 ms (0x faster)

[mutationObserverTest.html.zip](https://github.com/ampproject/amphtml/files/443033/mutationObserverTest.html.zip)

jridgewell commented on src/dom.js:
```diff
@@ -55,7 +56,14 @@ export function waitForChild(parent, checkFunc, callback) {
-  waitForChild(doc.documentElement, () => !!doc.body, callback);
+  // Prefer DocumentState.onBodyAvailable() to minimize creation of
+  // MutationObserver by waitForChild().
+  if (doc.defaultView) {
```
The document should always have a `defaultView`, right?

choumx replied:
I thought so too but I wasn't sure. [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Document/defaultView) mentions that it may be null, but this may never happen in AMP. I'll remove it.

jridgewell commented on src/document-state.js:
```diff
@@ -112,6 +116,31 @@ export class DocumentState {
+   */
+  onBodyAvailable(handler) {
+    const doc = this.document_;
+    if (!!doc.body) {
```
`!!` is unnecessary here.

choumx replied:
Ah I see conditionals already derive truthiness of its expr. Thanks!

jridgewell commented on src/document-state.js:
```diff
@@ -112,6 +116,31 @@ export class DocumentState {
+  /** @private */
+  onBodyAvailable_() {
+    this.bodyAvailableObservable_.fire();
+    this.bodyAvailableObservable_ = null;
```
We're going to have a memory leak here if anybody stores the `unlisten` returned by `#onBodyAvailable`. We'll need to add a `Observable#clear`.

choumx replied:
Added Observable#removeAll.

jridgewell commented:
LGTM.

cramforce commented on src/document-state.js:
```diff
@@ -65,6 +66,9 @@ export class DocumentState {
           this.boundOnVisibilityChanged_);
     }
+
+    /** @private @const {!Observable|null} */
```
`{?Observable}`

cramforce commented on src/document-state.js:
```diff
@@ -112,6 +116,32 @@ export class DocumentState {
+    if (!this.bodyAvailableObservable_) {
+      this.bodyAvailableObservable_ = new Observable();
+      waitForChild(doc.documentElement,
+                   () => !!doc.body,
```
nit: indent 4 instead of alignment with args on previous line.

choumx commented on src/document-state.js:
```diff
@@ -112,6 +116,32 @@ export class DocumentState {
   }
+
+  /**
+   * @param {function()} handler
```
Good idea. OOC, on iOS it's encouraged to invoke all callbacks async but I haven't seen that pattern in AMP. Is using setTimeout(f, 0) here bad form?

jridgewell replied:
[Z̧̡̩̞͕̘̳̗͖̱̉͋͛̆̄͋͆͊̕͠a̢̢̟͎̬͕̭̻͎͙̋͗̾̈́̂̀̊̽͑͘l̨̧̢̰̦̥͎̠̥̀͐̈̎̍̓͌̆̋͜͝g̢̡̘̮͙̞̟̭̹̳̀̅̓͗̐͑̀̕͝͝o̯̻͙̦̱̘̮̰̣̎̃͛͆̅̽͌̽͐̃͜](http://blog.izs.me/post/59142742143/designing-apis-for-asynchrony). Unfortunately, his dark highness found our codebase.

cramforce replied:
`setTimeout` can be pretty slow (up to event queue yield + 4ms). Promises avoid the problem (always async via a micro task). For a non-promise API `Promise.resolve().then(cb)` works.

choumx replied:
Good to know, thanks! Changing that here and in `waitForChild` would require fixing several unit tests. I'd be happy to make that change if you'd like.

choumx commented:
PTAL. Fixed the circular dependency by copying `waitForChild`'s implementation into `DocumentState`. This seems to be the least-bad way that minimizes API changes (and not many LOC of dupe). An alternative is to remove/replace `dom#waitForBody` with `DocumentState#onBodyAvailable` everywhere. Thanks!

jridgewell commented:
Why not just remove `waitForBody` entirely?

dvoytenko commented:
Just fyi, most likely we'll disallow the use of `dom.waitForBody` entirely and instead mandate the use of `ampdoc.whenBodyAvailable()` instead.

choumx commented:
> Why not just remove waitForBody entirely?

Re: Dima's comment, if `waitForBody` is going away soon anyways, perhaps it's not worthwhile to cause API churn twice by removing it now. This PR will effectively be a temporary optimization until Dima's change goes in. OTOH if that change is happening very soon, I can also just drop this PR.

dvoytenko commented:
@choumx Apologies, didn't review your PR until now. As it stands, this API may or may not conflict with the `ampdoc` API. The only way to tell would be to actually look at the uses of this API. I think the most likely resolution here is that (a) your changes are still needed, and (b) the uses will need to be migrated either to `ampdoc` API or your API on a case-by-case basis. Thus, it might be best to start from migrating the cases. For instance, in the context of `ampdoc` we should definitely use its API and when `ampdoc` is not available - your API should be used.

To give you an example, `custom-element.js` code is not tied to any document in particular and thus it makes sense to use a more global API. However, a `viewport-impl.js` is directly tied to an `ampdoc` and thus it should use that API instead. I'm currently working on migrating some of these services to `ampdoc` APIs anyway, so you can finish review of this PR and wait for me (or help me) finish the migration of ampdoc service. Then we can completely remove `dom.js` `waitForBody` API.

choumx commented:
Got it, thanks for the context. I undid the last change and instead reverted the implementation of `waitForBody`, deprecated it, and replaced a few usages with the new `DocumentState#onBodyAvailable`.

`waitForBody` is currently used in the following places:
- **extensions/amp-dynamic-css-classes/0.1/amp-dynamic-css-classes.js**
- **src/custom-element.js**
- src/runtime.js
- **src/style-installer.js**
- src/service/viewport-impl.js
- src/service/resources-impl.js

I changed the bolded files to use `onBodyAvailable` instead. I tried to infer which files were divorced from `AmpDoc`, but I could easily be wrong. PTAL!

jridgewell commented:
LGTM.

dvoytenko commented on src/style-installer.js:
```diff
@@ -159,12 +159,14 @@ export function makeBodyVisible(doc, opt_waitForExtensions) {
-    const perf = performanceFor(doc.defaultView);
+  const win = doc.defaultView;
+  const docState = documentStateFor(win);
+  docState.onBodyAvailable(() => {
```
This has to continue to use `waitForBody` from DOM because this code reacts to arrows where none of the services could have been installed yet.

dvoytenko replied:
Oh. Actually this is fine since, it looks like, we try to create `DocumentState` each time.

dvoytenko commented:
LGTM. Please rebase, we changed `makeBodyVisible` quite a bit.

choumx commented:
@dvoytenko @jridgewell Sorry for the delay -- this PR is ready to be merged!

jridgewell commented:
What are the remaining uses of `waitForBody`?

choumx commented:
The following files still use `waitForBody` or `waitForBodyPromise`:
- extensions/amp-experiment/0.1/amp-experiment.js
- extensions/amp-lightbox-viewer/0.1/service/lightbox-manager-discovery.js
- src/element-service.js
- src/runtime.js
- src/service/ampdoc-impl.js
- src/service/viewport-impl.js

choumx commented:
@dvoytenko @jridgewell Can this PR be merged? :)

jridgewell commented:
I'd say remove the remaining usages, but the PR LGTM.
